### ISSUE TITLE
report_submission(): a hashed, self-describing record of what a compilation cost

### ISSUE THREAD

**Issue by orgoca** [orgoca/CompileML#46]
Retention is a property of a dataset, a teacher and a depth budget — not of this library. The documentation site illustrates it with two runs of its own and says outright that two runs cannot answer the question.

Three hundred compiled models, each showing what compilation cost on somebody's real data, is a better argument for trying this library than any benchmark of ours. **That is what this is for: existence proof and range.** It is not a survey, and it cannot answer "what does compilation cost" in general — see *What this cannot claim*, which the site has to repeat wherever it shows a number.

This issue is the library half: producing a submission. Collecting and publishing it is a site concern, tracked at compileml-26a3504b#20.

## Not telemetry

Ruled out. The runtime's pitch is that it imports only the standard library and runs where the data already is, inside constrained environments. A library that phones home contradicts that, and most regulated institutions would block the call — so it would sample exactly the users least like the intended audience.

Nothing here sends anything. It writes a file the user reads, decides about, and submits by hand.

## What this cannot claim

A submitted corpus is self-selected twice over, and both need saying out loud
rather than discovered by a reader.

**Publication is voluntary.** People choose whether to submit, and the choice
is not independent of the result. The pull may be weaker than it first looks
— someone who gets 85% retention has a real incentive to publish and ask
whether that is normal, much as a suggestion box fills with complaints rather
than compliments — but "weaker than expected" is not "absent", and nothing
here measures which way it runs.

**Submission happens at the end of a successful compilation**, which is the
larger distortion and has nothing to do with incentives. A user who hits a
band-builder failure on a 5% base rate (#41), or gives up on a teacher that
will not distil, never reaches the function at all. The corpus therefore
excludes the hardest cases structurally — precisely the ones most worth
seeing. *Incomplete runs* below exists to close that gap.

So: publish distributions and counts, never a headline "retention is X%".
State the sample size and its self-selection next to every figure. Solicit
the runs that went badly as actively as the ones that went well.

## Capture widely, and capture now

Two reasons the payload should be generous rather than minimal.

**You cannot retroactively collect what you did not ask for.** The first fifty submissions are the ones that will most be wished richer, and they are exactly the ones that cannot be re-collected. Fields are cheap; missing history is not.

**A sweep is a curve, not a point.** This is what a one-row-per-submission design misses entirely. `sweep_whitebox` already returns, per configuration, Gini and retention against the teacher, Spearman agreement, whether attribution stayed exact, the quantized model's size and a measured explanation cost. If a user ran one, that is dozens of observations across trees × depth × alpha from a single submitter, on one dataset, with everything else held constant — precisely what a "trees versus retention" or "depth versus exactness cost" question needs, and unobtainable by pooling one-point submissions across incomparable datasets.

Including the sweep rows when present is the single highest-value field in the payload.

## Incomplete runs

A report should be generatable when compilation did **not** finish, carrying
`"status": "failed"`, whatever blocks completed, and the stage and error that
stopped it. A run that dies at `build_artifact` still has data, teacher and
compilation blocks worth having.

This is the highest-value part of the design and the least obvious. A registry
that shows where the library fails is more credible than one showing only
successes, not less — and it converts the blind spot above into a bug channel.
The two bugs found so far arrived because one person wrote an evaluation by
hand; a failed-run report is that same signal, from users who would never
write one.

## The payload

```json
{
  "report_version": 1,
  "status": "compiled",
  "compileml_version": "0.4.0",
  "created": "2026-09",
  "artifact_hash": "58274b10efd90f52...",
  "report_hash": "9c1f...",

  "data": {
    "n_rows": 1000000,
    "n_features": 42,
    "base_rate": 0.052,
    "use_case": "fraud",
    "holdout": { "method": "stratified", "test_fraction": 0.25 },
    "missing_share": 0.031
  },

  "teacher": {
    "family": "xgboost",
    "n_estimators": 400,
    "max_depth": 6,
    "learning_rate": 0.05,
    "sampling": "undersample",
    "tuned": true,
    "gini": 0.712,
    "ks": 0.541,
    "brier": 0.0388
  },

  "compilation": {
    "n_estimators": 30,
    "max_depth": 2,
    "learning_rate": 0.2,
    "loss": "squared_error",
    "alpha": 0.0,
    "monotone_constraints": { "declared": 5, "nonzero": 3 },
    "micro_scale": 1000000,
    "quantization_error_bound": 1.55e-05,
    "n_trees": 30,
    "artifact_kb": 104,
    "missing_policy": "baseline"
  },

  "bands": {
    "builder": "monotone_quantile_bands",
    "n_bands": 10,
    "merges": 0,
    "band_ordinal_gini": 0.691,
    "gini_gap": 0.008,
    "gini_gap_pct": 1.2,
    "worst_within_band_auc": 0.589,
    "verdicts": { "refinable": 2, "exhausted": 1, "inconclusive": 7 }
  },

  "outcome": {
    "gini": 0.699,
    "ks": 0.528,
    "brier": 0.0401,
    "retention_pct": 98.2,
    "spearman_vs_teacher": 0.977,
    "exact_attribution": true,
    "score_ms_median": 0.03,
    "explain_ms_median": 8.4
  },

  "reference": { "kind": "woe_logit", "gini": 0.664, "beats_reference": true },

  "validation": {
    "all_pass": true,
    "checks": {
      "1_integrity": "pass", "2_reconciliation": "pass", "3_fidelity": "skip",
      "4_band_properties": "pass", "5_semantic_monotonicity": "pass",
      "6_churn_baseline": "pass", "7_explainability_stability": "pass",
      "8_reason_coverage": "pass", "9_monotone_constraints": "pass",
      "10_reference_floor": "pass"
    }
  },

  "sweep": [
    { "n_estimators": 10, "max_depth": 1, "alpha": 0.0, "gini": 0.610,
      "retention_pct": 85.7, "spearman": 0.91, "exact_attribution": true,
      "model_kb": 12, "explain_ms": 3.1 }
  ],

  "notes": "~1M rows, heavily unbalanced at 5/95. XGBoost teacher, depth 6, trained with undersampling. Behavioural features only, no bureau data."
}
```

## What each block is for

**`teacher`** answers which families get compiled, and whether retention differs across them. Nothing in the record could say that before.

**`ks` and `brier` alongside `gini`, on both sides.** Gini is one view of one property. KS says where the separation sits; Brier says whether the calibration survived quantization — and calibration is the part most likely to be damaged by an integer isotonic table, so it is the metric most worth watching. Neither is exposed by the library today; `report_submission` would compute them, or small helpers get added.

**`bands`** turns a ladder into something comparable. `verdicts` in particular: how often a real ladder is still refinable says whether a default of ten bands is right, and no single dataset can answer that.

**`validation.checks` per check, not counts.** Which checks people skip is itself the finding — a high skip rate on `10_reference_floor` would say the reference model is too much work to fit, which is actionable.

**`compilation.quantization_error_bound` against `outcome.retention_pct`** is worth having on the record: is a looser bound associated with worse retention in practice, or is it slack? A self-selected corpus cannot settle that — it can only say whether the question deserves a designed experiment, which is a fair thing for it to do.

**`use_case`** stays a short controlled list — `credit_risk`, `fraud`, `churn`, `collections`, `marketing`, `other` — because the interesting statistic is conditional, and a pooled median across all of them is true of none.

## Two constraints this has to respect

**Every field is a disclosure surface, and richness trades against safety.** Hyperparameters and metrics are harmless. Combinations are not: `n_rows` plus `n_features` plus `base_rate` plus `use_case` could plausibly identify a portfolio in a small market. Hence `created` is a month rather than a timestamp, `base_rate` is rounded to three places, and the rule stays — **include what aggregates, exclude what identifies, and when they conflict, drop the field.** Never feature names, thresholds, leaf values, institution, or data.

**Capturing a field is not the same as publishing an analysis of it.** Correlations need many submissions across varied configurations; twenty reports will not support "trees versus efficiency" however rich each one is. Collect these now because they cannot be collected later, and let the site publish only what the sample supports — the same discipline already agreed for splitting by use case.

## Two things the hash does not do

**It does not authenticate.** `report_hash` over the canonical serialisation proves the document was not edited after generation. It does not prove the numbers are real. Same distinction the spec already draws for `artifact_hash`: integrity, not a signature. The published record has to say so.

What it does buy: tamper evidence, and deduplication on `artifact_hash` so the same artifact submitted twice is counted once.

It does **not** stop cherry-picking, and an earlier draft claimed it did. A rerun with different hyperparameters produces a different artifact and a different hash, so a submitter who tries five configurations and reports the best one passes deduplication untouched. Nothing in this design prevents that; the honest response is the one above — publish the spread rather than a headline, and treat the corpus as illustrative.

**It does not make the payload safe.** That is the constraint above — a decision per field, not a property of hashing.

## Scope note

Worth building before the site half, so the shape of a real report informs the ingestion schema rather than the reverse. Everything except `use_case`, `notes` and the teacher's own hyperparameters is already computable from what the caller has: `validate_artifact`, `band_efficiency`, `sweep_whitebox`, `fit_reference` and `quantization_error_bound` produce most of it today.

**Comment by orgoca** [comment]
Two additions from working through how the site should present this.

## Controlled vocabulary for the use case

Free text alone will not aggregate. The submission should carry a `use_case` field from a fixed, short list — something like `credit_risk`, `fraud`, `churn`, `collections`, `marketing`, `other` — alongside the free-text paragraph.

The reason is that the interesting statistic is almost certainly conditional. Retention on a 5/95 fraud panel is a different question from retention on a bureau scorecard: different base rates, different teacher families, different depth budgets in practice. A single pooled median across all of them would average two answers and be true of neither. With a controlled field the record can be split, and each split says something.

Free text stays, because it carries what the enum cannot — imbalance, sampling strategy, feature provenance, whether the teacher was tuned. The enum makes the table groupable; the paragraph makes an individual row interpretable.

Keep the list short and add to it reluctantly. A twenty-value enum with three entries per value aggregates no better than free text.

## Selection bias needs to be stated on the page, not in a footnote

Nobody reports a disappointment. A compilation that gave up eight points of Gini is exactly the entry this record most needs and is least likely to receive, and no amount of curation fixes that — it is a property of who chooses to submit.

So the published table has to say plainly that it is the favourable edge of what compilation costs rather than the middle of it, and the template should say outright that a bad result is the more valuable submission. That will not fully correct the bias, but stating it is the difference between a record and an advertisement.

The site now carries that caveat in the second paragraph of the page, above the numbers rather than below them.

## Where the site stands

`/benchmarks` is now "What compilation has cost so far", written as a record that currently holds two runs — the seeded synthetic one and the public UCI panel — and is explicit that two runs of our own can only illustrate the question rather than answer it.

**Comment by orgoca** [comment]
## Trust model — decided, so nobody builds for the wrong one

**Assume good faith. No signing, no attestation, no anti-abuse machinery.**

Someone spending their time assembling a report is overwhelmingly not trolling, and the failure mode if one slips through is a single wrong row in a curated table, removed by hand when spotted. That is a cheap failure. Building cryptographic provenance to prevent it would be expensive, would not work anyway — a signature proves who sent a number, never that the number is true — and would push the payload toward identifying its author, which is the opposite of what this needs.

So `report_hash` stays what it is: a canonical hash for **tamper evidence and deduplication**, not verification. The published record will say plainly that entries are self-reported and not independently verified. That sentence does more honest work than any signature would.

Concretely, do not implement: signing, key management, identity, rate limiting, submission endpoints, or any network call.

## Notes for whoever picks this up

Not prescriptive about design — just what I know from the site side, so none of it has to be rediscovered.

**Reuse rather than recompute.** `runtime.canonical_hash` already does canonical serialisation and hashing; `report_hash` should be that function over the payload with the field absent, matching how `artifact_hash` works. `validate_artifact` returns `{"all_pass", "checks": {...}}` with a `skipped` flag per check, so `checks_passed` / `checks_skipped` come from counting it rather than re-running anything. `reference.fit_reference` and `reference_gini` give the floor fields.

**It is compile-side, not runtime.** The stdlib-only constraint does not apply — numpy and scikit-learn are fine. Do not put it under `compileml/runtime/`, or the import-purity test will fail.

**Worth having tests for:** the payload round-trips and its hash verifies; editing any field invalidates the hash; two runs on the same artifact produce the same `artifact_hash` so dedup works; `use_case` outside the controlled list is rejected rather than passed through; and — the one that matters most — **the emitted document contains no feature names**, since that is the disclosure rule and a test is the only thing that keeps it true after a refactor. Nothing here should need the network.

**A CLI surface would help adoption** (`compileml report --artifact … --csv … --y-col …`), since the audience for submitting is broader than the audience writing Python against the API. `docs/reference/cli.md` documents commands at that granularity already.

**Docs:** a `:::` directive in `docs/reference/api.md` picks it up automatically, and the how-to section is the natural home for a page on submitting. A good docstring gets it into `help()`, the mkdocs site and compileml.com in one edit.

## Downstream

The documentation site cannot ingest anything until this is released and its pin moves — `pipeline/pin.json` deliberately refuses a pin that is not a released tag. Site-side ingestion and summary statistics are tracked at compileml-26a3504b#20 and are explicitly sequenced after this, so that the shape of a real report informs the schema rather than the reverse.

**Comment by orgoca** [comment]
Revised after review. Three changes, all narrowing what this claims rather than what it captures.

**The purpose is existence proof, not inference.** The opening argued that a corpus of other people's compilations *would* answer what two runs cannot. It would not — it is self-selected, and answering that question is not what makes it worth building. Three hundred real compilations with their costs on the record is a better reason to try the library than any benchmark of mine, and that stands on its own without pretending to be a survey. The inferential promise attached to `quantization_error_bound` is softened to match: a corpus like this can say whether a question deserves a designed experiment, not settle it.

**A new section names both biases outright.** Publication is voluntary, and the pull may be weaker than it looks — someone at 85% retention has a real incentive to publish and ask whether that is normal, much as a suggestion box fills with complaints rather than compliments. But the larger distortion has nothing to do with incentives: `report_submission()` runs at the end of a *successful* compilation, so anyone who hit #41 on a 5% base rate, or gave up on a teacher that would not distil, never reaches it. The corpus structurally excludes the hardest cases.

**Hence incomplete runs are now first-class.** A report should be generatable with `"status": "failed"`, carrying whatever completed plus the stage that stopped it. This is the part I would defend hardest: a registry showing where the library fails is more credible than one showing only successes, and it turns the blind spot into a bug channel. Both bugs found so far surfaced because one person wrote an evaluation by hand — a failed-run report is that same signal from users who would never write one.

Also corrected: the claim that deduplication on `artifact_hash` stops resubmission after a favourable rerun. It does not. A rerun with different hyperparameters is a different artifact with a different hash, so someone who tries five configurations and reports the best passes deduplication untouched. Nothing here prevents cherry-picking; the honest response is to publish the spread and treat the corpus as illustrative.

Still not scheduled. #41 breaks the documented path on low-base-rate data, the runtime API is undocumented, and #10 is the larger adoption item — all ahead of this.

**Comment by orgoca** [comment]
**Parked, deliberately.** Not abandoned and not blocked — there is simply no queue of submissions waiting on it, and the design benefits from seeing a real report before the schema is fixed.

The spec above is complete enough to pick up cold whenever it becomes worth building: payload, the trust model (assume good faith, no signing), the disclosure rules, and the reasoning for including the sweep rows.

Companion issue on the documentation site — ingestion, curation and summary statistics — is parked alongside it.
